File: leetcode_cn/49.py
from collections import defaultdict


class D:
    def __init__(self, d: dict[str, int]) -> None:
        self.d = d

    def __hash__(self) -> int:
        h = 0
        for k, v in sorted(self.d.items(), key=lambda x: x[0]):
            h += (1 << ((ord(k) - ord("a")) * 7)) * v
        return h

    # def print(self) -> None:
    #     print(self.d)


class Solution:
    def preprocess(self, s: str) -> dict[str, int]:
        d: dict[str, int] = defaultdict(int)
        for i in s:
            d[i] += 1
        return d

    def groupAnagrams(self, strs: list[str]) -> list[list[str]]:
        dct: dict[int, list[str]] = defaultdict(list)
        for s in strs:
            tmp = D(self.preprocess(s))
            # tmp.print()
            # print(hash(tmp))
            dct[hash(tmp)].append(s)
        return list(dct.values())


if __name__ == "__main__":
    s = Solution()
    print(s.groupAnagrams(["eat", "tea", "tan", "ate", "nat", "bat"]))
    print(s.groupAnagrams(["ddddddddddg", "dgggggggggg"]))

File: leetcode_cn/49.rs
use std::collections::{BTreeMap, HashMap};

struct Solution;

impl Solution {
    pub fn preprocess(s: &str) -> BTreeMap<char, i32> {
        let mut d: BTreeMap<char, i32> = BTreeMap::new();
        for i in s.chars() {
            *d.entry(i).or_insert(0) += 1;
        }
        d
    }
    pub fn group_anagrams(strs: Vec<String>) -> Vec<Vec<String>> {
        let mut dct: HashMap<BTreeMap<char, i32>, Vec<String>> = HashMap::new();
        for s in strs.into_iter() {
            let tmp = Self::preprocess(&s);
            dct.entry(tmp).or_insert(vec![]).push(s);
        }
        dct.into_values().collect()
    }
}

fn main() {
    println!(
        "{:?}",
        Solution::group_anagrams(vec![
            "eat".to_string(),
            "tea".to_string(),
            "tan".to_string(),
            "ate".to_string(),
            "nat".to_string(),
            "bat".to_string()
        ])
    );
    println!(
        "{:?}",
        Solution::group_anagrams(vec!["ddddddddddg".to_string(), "dgggggggggg".to_string()])
    );
}
